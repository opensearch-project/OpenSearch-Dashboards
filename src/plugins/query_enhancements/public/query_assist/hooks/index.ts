/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export * from './use_generate';
export * from './use_query_assist';
