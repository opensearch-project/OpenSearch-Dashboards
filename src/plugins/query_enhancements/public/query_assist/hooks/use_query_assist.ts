/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
import React from 'react';
import { BehaviorSubject } from 'rxjs';

export interface QueryAssistContextValue {
  question: string;
  question$: BehaviorSubject<string>;
  updateQuestion: (question: string) => void;
  isQueryAssistCollapsed: boolean;
  updateIsQueryAssistCollapsed: (isCollapsed: boolean) => void;
}
export const QueryAssistContext = React.createContext<QueryAssistContextValue>(
  {} as QueryAssistContextValue
);
export const useQueryAssist = () => React.useContext(QueryAssistContext);
