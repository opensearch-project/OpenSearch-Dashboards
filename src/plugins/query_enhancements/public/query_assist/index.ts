/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export { createQueryAssistExtension } from './utils';
