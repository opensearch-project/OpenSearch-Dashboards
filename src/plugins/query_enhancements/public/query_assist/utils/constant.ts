/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export const DATA2SUMMARY_AGENT_CONFIG_ID = 'os_data2summary';
