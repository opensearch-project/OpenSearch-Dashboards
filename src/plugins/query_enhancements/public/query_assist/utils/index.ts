/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export * from './create_extension';
export * from './errors';
export * from './get_persisted_log';
