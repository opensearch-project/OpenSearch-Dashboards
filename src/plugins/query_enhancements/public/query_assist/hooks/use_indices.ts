/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { CatIndicesResponse } from '@opensearch-project/opensearch/api/types';
import { Reducer, useEffect, useReducer, useState } from 'react';
import { IDataPluginServices } from '../../../../data/public';
import { useOpenSearchDashboards } from '../../../../opensearch_dashboards_react/public';

interface State<T> {
  data?: T;
  loading: boolean;
  error?: Error;
}

type Action<T> =
  | { type: 'request' }
  | { type: 'success'; payload: State<T>['data'] }
  | { type: 'failure'; error: NonNullable<State<T>['error']> };

// TODO use instantiation expressions when typescript is upgraded to >= 4.7
type GenericReducer<T = any> = Reducer<State<T>, Action<T>>;
export const genericReducer: GenericReducer = (state, action) => {
  switch (action.type) {
    case 'request':
      return { data: state.data, loading: true };
    case 'success':
      return { loading: false, data: action.payload };
    case 'failure':
      return { loading: false, error: action.error };
    default:
      return state;
  }
};

export const useIndices = (dataSourceId: string | undefined) => {
  const reducer: GenericReducer<string[]> = genericReducer;
  const [state, dispatch] = useReducer(reducer, { loading: false });
  const [refresh, setRefresh] = useState({});
  const { services } = useOpenSearchDashboards<IDataPluginServices>();

  useEffect(() => {
    const abortController = new AbortController();
    dispatch({ type: 'request' });
    services.http
      .post('/api/console/proxy', {
        query: { path: '_cat/indices?format=json', method: 'GET', dataSourceId },
        signal: abortController.signal,
      })
      .then((payload: CatIndicesResponse) =>
        dispatch({
          type: 'success',
          payload: payload
            .filter((meta) => meta.index && !meta.index.startsWith('.'))
            .map((meta) => meta.index!),
        })
      )
      .catch((error) => dispatch({ type: 'failure', error }));

    return () => abortController.abort();
  }, [refresh, services.http, dataSourceId]);

  return { ...state, refresh: () => setRefresh({}) };
};

export const useIndexPatterns = () => {
  const reducer: GenericReducer<string[]> = genericReducer;
  const [state, dispatch] = useReducer(reducer, { loading: false });
  const [refresh, setRefresh] = useState({});
  const { services } = useOpenSearchDashboards<IDataPluginServices>();

  useEffect(() => {
    let abort = false;
    dispatch({ type: 'request' });

    services.data.indexPatterns
      .getTitles()
      .then((payload) => {
        if (!abort)
          dispatch({
            type: 'success',
            // temporary solution does not support index patterns from other data sources
            payload: payload.filter((title) => !title.includes('::')),
          });
      })
      .catch((error) => {
        if (!abort) dispatch({ type: 'failure', error });
      });

    return () => {
      abort = true;
    };
  }, [refresh, services.data.indexPatterns]);

  return { ...state, refresh: () => setRefresh({}) };
};
