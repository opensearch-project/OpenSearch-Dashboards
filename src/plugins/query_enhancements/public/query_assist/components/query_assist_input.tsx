/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiFieldText, EuiIcon, EuiOutsideClickDetector, EuiPortal } from '@elastic/eui';
import { i18n } from '@osd/i18n';
import React, { useMemo, useState } from 'react';
import { PersistedLog, QuerySuggestionTypes } from '../../../../data/public';
import assistantMark from '../../assets/sparkle_mark.svg';
import { getData } from '../../services';
import { AgentError } from '../utils';
import { WarningBadge } from './warning_badge';

interface QueryAssistInputProps {
  inputRef: React.RefObject<HTMLInputElement>;
  persistedLog: PersistedLog;
  isDisabled: boolean;
  initialValue?: string;
  selectedIndex?: string;
  previousQuestion?: string;
  error?: AgentError;
}

export const QueryAssistInput: React.FC<QueryAssistInputProps> = (props) => {
  const {
    ui: { SuggestionsComponent },
  } = getData();
  const [isSuggestionsVisible, setIsSuggestionsVisible] = useState(false);
  const [suggestionIndex, setSuggestionIndex] = useState<number | null>(null);
  const [value, setValue] = useState(props.initialValue ?? '');

  const sampleDataSuggestions = useMemo(() => {
    switch (props.selectedIndex) {
      case 'opensearch_dashboards_sample_data_ecommerce':
        return [
          'How many unique customers placed orders this week?',
          'Count the number of orders grouped by manufacturer and category',
          'find customers with first names like Eddie',
        ];

      case 'opensearch_dashboards_sample_data_logs':
        return [
          'Are there any errors in my logs?',
          'How many requests were there grouped by response code last week?',
          "What's the average request size by week?",
        ];

      case 'opensearch_dashboards_sample_data_flights':
        return [
          'how many flights were there this week grouped by destination country?',
          'what were the longest flight delays this week?',
          'what carriers have the furthest flights?',
        ];

      default:
        return [];
    }
  }, [props.selectedIndex]);

  const suggestions = useMemo(() => {
    if (!props.persistedLog) return [];
    return props.persistedLog
      .get()
      .concat(sampleDataSuggestions)
      .filter(
        (suggestion, i, array) => array.indexOf(suggestion) === i && suggestion.includes(value)
      )
      .map((suggestion) => ({
        type: QuerySuggestionTypes.RecentSearch,
        text: suggestion,
        start: 0,
        end: value.length,
      }));
  }, [props.persistedLog, value, sampleDataSuggestions]);

  return (
    <EuiOutsideClickDetector onOutsideClick={() => setIsSuggestionsVisible(false)}>
      <div className="queryAssist__inputWrapper">
        <EuiFieldText
          className="queryAssist__input"
          data-test-subj="query-assist-input-field-text"
          inputRef={props.inputRef}
          value={value}
          disabled={props.isDisabled}
          onClick={() => setIsSuggestionsVisible(true)}
          onChange={(e) => setValue(e.target.value)}
          onKeyDown={() => setIsSuggestionsVisible(true)}
          placeholder={
            props.previousQuestion ||
            (props.selectedIndex
              ? i18n.translate('queryEnhancements.queryAssist.input.placeholderWithIndex', {
                  defaultMessage:
                    'Ask a natural language question about {selectedIndex} to generate a query',
                  values: { selectedIndex: props.selectedIndex },
                })
              : i18n.translate('queryEnhancements.queryAssist.input.placeholderWithoutIndex', {
                  defaultMessage: 'Select an index to ask a question',
                }))
          }
          prepend={<EuiIcon type={assistantMark} />}
          append={<WarningBadge error={props.error} />}
          fullWidth
        />
        <EuiPortal>
          <SuggestionsComponent
            show={isSuggestionsVisible}
            suggestions={suggestions}
            index={suggestionIndex}
            onClick={(suggestion) => {
              if (!props.inputRef.current) return;
              setValue(suggestion.text);
              setIsSuggestionsVisible(false);
              setSuggestionIndex(null);
              props.inputRef.current.focus();
            }}
            onMouseEnter={(i) => setSuggestionIndex(i)}
            loadMore={() => {}}
            queryBarRect={props.inputRef.current?.getBoundingClientRect()}
            size="s"
          />
        </EuiPortal>
      </div>
    </EuiOutsideClickDetector>
  );
};
