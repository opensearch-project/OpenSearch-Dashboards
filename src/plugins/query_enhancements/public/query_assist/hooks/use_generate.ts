/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { useEffect, useRef, useState } from 'react';
import { IDataPluginServices } from '../../../../data/public';
import { useOpenSearchDashboards } from '../../../../opensearch_dashboards_react/public';
import { API } from '../../../common';
import { QueryAssistParameters, QueryAssistResponse } from '../../../common/query_assist';
import { formatError } from '../utils';

export const useGenerateQuery = () => {
  const mounted = useRef(false);
  const [loading, setLoading] = useState(false);
  const abortControllerRef = useRef<AbortController>();
  const { services } = useOpenSearchDashboards<IDataPluginServices>();

  useEffect(() => {
    mounted.current = true;
    return () => {
      mounted.current = false;
      if (abortControllerRef.current) {
        abortControllerRef.current.abort();
        abortControllerRef.current = undefined;
      }
    };
  }, []);

  const generateQuery = async (
    params: QueryAssistParameters
  ): Promise<{ response?: QueryAssistResponse; error?: Error }> => {
    abortControllerRef.current?.abort();
    abortControllerRef.current = new AbortController();
    setLoading(true);
    try {
      const response = await services.http.post<QueryAssistResponse>(API.QUERY_ASSIST.GENERATE, {
        body: JSON.stringify(params),
        signal: abortControllerRef.current?.signal,
      });
      if (mounted.current) return { response };
    } catch (error) {
      if (mounted.current) return { error: formatError(error) };
    } finally {
      if (mounted.current) setLoading(false);
    }
    return {};
  };

  return { generateQuery, loading, abortControllerRef };
};
