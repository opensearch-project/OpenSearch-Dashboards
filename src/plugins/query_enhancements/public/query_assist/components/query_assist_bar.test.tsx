/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import React, { ComponentProps, PropsWithChildren } from 'react';
import { IntlProvider } from 'react-intl';
import { of } from 'rxjs';
import { QueryAssistBar } from '.';
import { notificationServiceMock, uiSettingsServiceMock } from '../../../../../core/public/mocks';
import { DataStorage } from '../../../../data/common';
import { QueryEditorExtensionDependencies, QueryStringContract } from '../../../../data/public';
import { dataPluginMock } from '../../../../data/public/mocks';
import { useOpenSearchDashboards } from '../../../../opensearch_dashboards_react/public';
import { setData, setStorage } from '../../services';
import { useGenerateQuery } from '../hooks';
import { AgentError, ProhibitedQueryError } from '../utils';
import { QueryAssistInput } from './query_assist_input';

jest.mock('../../../../opensearch_dashboards_react/public', () => ({
  useOpenSearchDashboards: jest.fn(),
  withOpenSearchDashboards: jest.fn((component: React.Component) => component),
}));

jest.mock('../hooks', () => ({
  useGenerateQuery: jest.fn().mockReturnValue({ generateQuery: jest.fn(), loading: false }),
}));

jest.mock('./query_assist_input', () => ({
  QueryAssistInput: ({ inputRef, error }: ComponentProps<typeof QueryAssistInput>) => (
    <>
      <input ref={inputRef} />
      <div>{JSON.stringify(error)}</div>
    </>
  ),
}));

const dataMock = dataPluginMock.createStartContract(true);
const queryStringMock = dataMock.query.queryString as jest.Mocked<QueryStringContract>;
const uiSettingsMock = uiSettingsServiceMock.createStartContract();
const notificationsMock = notificationServiceMock.createStartContract();

setData(dataMock);
setStorage(new DataStorage(window.localStorage, 'mock-prefix'));

const dependencies: QueryEditorExtensionDependencies = {
  language: 'PPL',
  onSelectLanguage: jest.fn(),
  isCollapsed: false,
  setIsCollapsed: jest.fn(),
};

type Props = ComponentProps<typeof QueryAssistBar>;

const IntlWrapper = ({ children }: PropsWithChildren<unknown>) => (
  <IntlProvider locale="en">{children}</IntlProvider>
);

const renderQueryAssistBar = (overrideProps: Partial<Props> = {}) => {
  const props: Props = Object.assign<Props, Partial<Props>>({ dependencies }, overrideProps);
  const component = render(<QueryAssistBar {...props} />, {
    wrapper: IntlWrapper,
  });
  return { component, props: props as jest.MockedObjectDeep<Props> };
};

describe('QueryAssistBar', () => {
  beforeEach(() => {
    (useOpenSearchDashboards as jest.Mock).mockReturnValue({
      services: {
        data: dataMock,
        uiSettings: uiSettingsMock,
        notifications: notificationsMock,
      },
    });
  });
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders null if collapsed', () => {
    const { component } = renderQueryAssistBar({
      dependencies: { ...dependencies, isCollapsed: true },
    });
    expect(component.container).toBeEmptyDOMElement();
  });

  it('matches snapshot', () => {
    const { component } = renderQueryAssistBar();
    expect(component.container).toMatchSnapshot();
  });

  it('displays callout when query input is empty on submit', async () => {
    renderQueryAssistBar();

    fireEvent.click(screen.getByRole('button'));

    await waitFor(() => {
      expect(screen.getByTestId('query-assist-empty-query-callout')).toBeInTheDocument();
    });
  });

  it('displays callout when dataset is not selected on submit', async () => {
    queryStringMock.getQuery.mockReturnValueOnce({ query: '', language: 'kuery' });
    queryStringMock.getUpdates$.mockReturnValueOnce(of({ query: '', language: 'kuery' }));
    renderQueryAssistBar();

    fireEvent.change(screen.getByRole('textbox'), { target: { value: 'test query' } });
    fireEvent.click(screen.getByRole('button'));

    await waitFor(() => {
      expect(screen.getByTestId('query-assist-empty-index-callout')).toBeInTheDocument();
    });
  });

  it('displays callout for guardrail errors', async () => {
    const generateQueryMock = jest.fn().mockResolvedValue({ error: new ProhibitedQueryError() });
    (useGenerateQuery as jest.Mock).mockReturnValue({
      generateQuery: generateQueryMock,
      loading: false,
    });

    renderQueryAssistBar();

    fireEvent.change(screen.getByRole('textbox'), { target: { value: 'test query' } });
    fireEvent.click(screen.getByRole('button'));

    await waitFor(() => {
      expect(screen.getByTestId('query-assist-guard-callout')).toBeInTheDocument();
    });
  });

  it('passes agent errors to input', async () => {
    const generateQueryMock = jest.fn().mockResolvedValue({
      error: new AgentError({
        error: { type: 'mock-type', reason: 'mock-reason', details: 'mock-details' },
        status: 303,
      }),
    });
    (useGenerateQuery as jest.Mock).mockReturnValue({
      generateQuery: generateQueryMock,
      loading: false,
    });

    renderQueryAssistBar();

    fireEvent.change(screen.getByRole('textbox'), { target: { value: 'test query' } });
    fireEvent.click(screen.getByRole('button'));

    await waitFor(() => {
      expect(screen.getByText(/mock-reason/)).toBeInTheDocument();
    });
  });

  it('displays toast for other unknown errors', async () => {
    const mockError = new Error('mock-error');
    const generateQueryMock = jest.fn().mockResolvedValue({
      error: mockError,
    });
    (useGenerateQuery as jest.Mock).mockReturnValue({
      generateQuery: generateQueryMock,
      loading: false,
    });

    renderQueryAssistBar();

    fireEvent.change(screen.getByRole('textbox'), { target: { value: 'test query' } });
    fireEvent.click(screen.getByRole('button'));

    await waitFor(() => {
      expect(notificationsMock.toasts.addError).toHaveBeenCalledWith(mockError, {
        title: 'Failed to generate results',
      });
    });
  });

  it('submits a valid query and updates services', async () => {
    const generateQueryMock = jest
      .fn()
      .mockResolvedValue({ response: { query: 'generated query' } });
    (useGenerateQuery as jest.Mock).mockReturnValue({
      generateQuery: generateQueryMock,
      loading: false,
    });

    renderQueryAssistBar();

    fireEvent.change(screen.getByRole('textbox'), { target: { value: 'test query' } });
    fireEvent.click(screen.getByRole('button'));

    await waitFor(() => {
      expect(generateQueryMock).toHaveBeenCalledWith({
        question: 'test query',
        index: 'Default Index Pattern',
        language: 'PPL',
        dataSourceId: 'mock-data-source-id',
      });
    });

    expect(queryStringMock.setQuery).toHaveBeenCalledWith({
      dataset: {
        dataSource: {
          id: 'mock-data-source-id',
          title: 'Default Data Source',
          type: 'OpenSearch',
        },
        id: 'default-index-pattern',
        timeFieldName: '@timestamp',
        title: 'Default Index Pattern',
        type: 'INDEX_PATTERN',
      },
      language: 'PPL',
      query: 'generated query',
    });
    expect(screen.getByTestId('query-assist-query-generated-callout')).toBeInTheDocument();
  });
});
