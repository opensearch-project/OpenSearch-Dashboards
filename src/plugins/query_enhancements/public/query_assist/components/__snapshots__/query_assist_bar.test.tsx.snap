// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryAssistBar matches snapshot 1`] = `
<div>
  <form
    class="euiForm queryAssist queryAssist__form"
  >
    <div
      class="euiFormRow euiFormRow--fullWidth"
      id="generated-id-row"
    >
      <div
        class="euiFormRow__fieldWrapper"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterExtraSmall euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow"
          id="generated-id"
        >
          <div
            class="euiFlexItem"
          >
            <input />
            <div />
          </div>
          <div
            class="euiFlexItem euiFlexItem--flexGrowZero"
          >
            <button
              aria-label="Submit question to query assistant"
              class="euiButtonIcon euiButtonIcon--primary euiButtonIcon--small"
              data-test-subj="query-assist-submit-button"
              type="submit"
            >
              <span
                aria-hidden="true"
                class="euiButtonIcon__icon"
                color="inherit"
                data-euiicon-type="returnKey"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </form>
</div>
`;
